### PR TITLE
fix: making auth and firestore observable compatible

### PR DESCRIPTION
**Pull request by zhigang1992**
### Description

Official firebase-js-sdk support observer as a form of callback, we have partial support on Firestore's onSnapshot.

But it wasn't 100% inline with js-sdk, for example. previously, we were just passing observer.next directly. which will cause problem if that function has reference to observer via this. Adding bind(this) should do the trick

### Related issues

https://github.com/invertase/react-native-firebase/issues/2066

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

File: packages/auth/lib/index.d.ts [hunk 1]
```diff
@@ -1634,6 +1634,8 @@ export namespace FirebaseAuthTypes {
   }
 }
 
+type CallbackOrObserver<T extends (...args: any[]) => any> = T | { next: T };
```
Comment (zhigang1992) [inline packages/auth/lib/index.d.ts:1637]: No very sure on where to put this type

File: packages/auth/lib/index.js [hunk 2]
```diff
@@ -129,7 +129,14 @@ class FirebaseAuthModule extends FirebaseModule {
     };
   }
 
-  onAuthStateChanged(listener) {
+  _parseListener(listenerOrObserver) {
+    return typeof listenerOrObserver === 'object'
+      ? listenerOrObserver.next.bind(listenerOrObserver)
+      : listenerOrObserver;
+  }
```
Comment (zhigang1992) [inline packages/auth/lib/index.js:136]: also not sure where to put this utils function.

**Comment by mikehardy** [comment]
Hi there!

E2E test failure, possibly related? I haven't reviewed the PR yet - but was working on CI yesterday and wanted to see if the CI failure was false negative or real, so it's all I've examined so far:

```
  1) firestore().doc().onSnapshot()
       throws if SnapshotListenerOptions is invalid:
     Error: Did not throw an Error.
      at Context.<anonymous> (/Users/runner/work/react-native-firebase/react-native-firebase/packages/firestore/e2e/DocumentReference/onSnapshot.e2e.js:302:29)
      at processImmediate (internal/timers.js:456:21)
```

**Comment by zhigang1992** [comment]
> Hi there!
> 
> E2E test failure, possibly related? I haven't reviewed the PR yet - but was working on CI yesterday and wanted to see if the CI failure was false negative or real, so it's all I've examined so far:
> 
> ```
>   1) firestore().doc().onSnapshot()
>        throws if SnapshotListenerOptions is invalid:
>      Error: Did not throw an Error.
>       at Context.<anonymous> (/Users/runner/work/react-native-firebase/react-native-firebase/packages/firestore/e2e/DocumentReference/onSnapshot.e2e.js:302:29)
>       at processImmediate (internal/timers.js:456:21)
> ```

Yeah, will fix it ASAP

**Comment by Salakar** [comment]
Nice, thanks for sending this over

**Comment by mikehardy** [comment]
@zhigang1992 - strange - I'm not sure if this is related to your change or not, but the current CI failure might be?

```

  1) firestore().collection().onSnapshot()
       unsubscribes from further updates:
     AssertionError: expected Function { name: '' } to be called thrice but was called 4 times
    spy([object Object], null) at handleSuccess (http://localhost:8081/index.bundle?platform=android&dev=true&minify=false&sourceMapURL=true:202068:11)
    spy([object Object], null) at handleSuccess (http://localhost:8081/index.bundle?platform=android&dev=true&minify=false&sourceMapURL=true:202068:11)
    spy([object Object], null) at handleSuccess (http://localhost:8081/index.bundle?platform=android&dev=true&minify=false&sourceMapURL=true:202068:11)
    spy([object Object], null) at handleSuccess (http://localhost:8081/index.bundle?platform=android&dev=true&minify=false&sourceMapURL=true:202068:11)
  
```

**Comment by mikehardy** [comment]
Same thing happened on #1039 - I think it's a flaky test so I'm re-running the job and linking it to #4058 for attention

**Comment by mikehardy** [comment]
Working through a merge of all PRs that are ready, will take a while (probably in to tomorrow) to finish that, then I'll do the release. Thanks again @zhigang1992 !

**Pull request merged**